### DISCUSSION TITLE
Let's Explore: IDL (Interface Description Language)

### DISCUSSION THREAD

**Discussion by humbletim** [omigroup/omigroup#24] [R&D]
Collaborative research and development around using IDLs to capture and prototype interoperable data exchanges between platforms.

- initial experiments repo: https://github.com/XRFoundation/open-metaverse-interface @shawticus

**Comment by humbletim** [comment]
brainstorming data elements to consider, @madjin shared a link the other day to [VRChat's API](https://vrchatapi.github.io/#).

similar to the draft experimental IDL schema, VRChat interfaces also differentiate (slightly) between the idea of the "current user" and other "user":
- User -- https://vrchatapi.github.io/#/Objects/User?id=user-object
- Base User -- https://vrchatapi.github.io/#/Objects/User?id=current-user-object

**Reply by humbletim** [reply]
as another reference point here is a rough impression of how Webaverse might ponder "user" schema:
```js
userObject = {
    name,
    avatar: { id, name, ext, preview, url },
    loadout,
    inventory,
    homespace: { id, name, ext, preview, url },
 }
  ```
  - see userObject - https://github.com/webaverse/app/blob/1c8fd37/login.js#L54-L74

**Comment by lalalune** [comment]
Loadout is an example of where we need some sort of flexible, possibly namespaced data. Some people will want loadout and some people won't.

If we want these concepts to be part of the spec, we could create archetypal interfaces that people can build on-- kind of like how ERC-721 can be implemented for all kinds of different features in NFTs. That way verses who choose to implement loadout can interface with the existing spec, if it's close enough to their own, or fork and create a new version (similar to IPSME).

**Reply by humbletim** [reply]
That sounds reasonable -- one place archetypal interfaces and namespaced extensions seemed to work fairly well is with browser vendors eventually standardizing on CSS3.  Like, starting from CSS1, browsers began implementing the core spec while also experimenting with [vendor-specific prefixes](https://developer.mozilla.org/en-US/docs/Glossary/Vendor_Prefix).

Seems like a delicate balance to maintain, but where extensions can be expressed as plain text (ie: not as proprietary binary blobs), it seems possible to view namespaced extensions as similar to "proposals" for new features as well as invitations for other vendors to offer "counter-proposals" in their own dialect/prefixes for consideration.
